esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -0,0 +1,204 @@
+            throw MakeStringException(-1, "CustomResourcesBaseDN must be specified (ex. 'ou=smc,ou=espservices,ou=hpcc,dc=myLDAPDomain,dc=com')");
+        if (!req.getWorkunitsMode()==2 && isEmptyString(req.getCustomWorkunitsBaseDN()))
+            throw MakeStringException(-1, "CustomWorkunitsBaseDN must be specified (ex. 'ou=workunits,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        // Create OU names
+
+        StringBuffer respFilesBaseDN, respGroupsBaseDN, respUsersBaseDN, respResourcesBaseDN, respWorkunitsBaseDN;
+        formatOUname(respFilesBaseDN, req.getEnvName(), req.getFilesMode(), sharedFilesBaseDN.str(), req.getCustomFilesBaseDN(), "ou=files");
+        formatOUname(respGroupsBaseDN, req.getEnvName(), req.getGroupsMode(), sharedGroupsBaseDN.str(), req.getCustomGroupsBaseDN(), "ou=groups");
+        formatOUname(respUsersBaseDN, req.getEnvName(), req.getUsersMode(), sharedUsersBaseDN.str(), req.getCustomUsersBaseDN(), "ou=users");
+        formatOUname(respResourcesBaseDN, req.getEnvName(), req.getResourcesMode(), sharedResourcesBaseDN.str(), req.getCustomResourcesBaseDN(), "ou=smc,ou=espservices");
+        formatOUname(respWorkunitsBaseDN, req.getEnvName(), req.getWorkunitsMode(), sharedWorkunitsBaseDN.str(), req.getCustomWorkunitsBaseDN(), "ou=workunits");
+
+        // Create HPCC Admin Username
+
+        VStringBuffer respAdminUser("admin_%s", req.getEnvName());
+
+
+        //Create LDAP resources OU Hierarchy
+
+        if (req.getCreateLDAPEnvironment())
+        {
+            VStringBuffer description("%s, created by %s", req.getEnvDescription(), req.getEnvOwnerName());
+            try { secmgr->createLdapBasedn(nullptr, respFilesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
+            catch(...) {}
+            try { secmgr->createLdapBasedn(nullptr, respGroupsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
+            catch(...) {}
+            try { secmgr->createLdapBasedn(nullptr, respUsersBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
+            catch(...) {}
+            try { secmgr->createLdapBasedn(nullptr, respResourcesBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
+            catch(...) {}
+            try { secmgr->createLdapBasedn(nullptr, respWorkunitsBaseDN.str(), PT_ADMINISTRATORS_ONLY, description.str()); }
+            catch(...) {}
+
+            // NOTE!  HPCC Admin group and HPCC Admin user will be
+            // created when ECLWatch is started in Admin mode
+        }
+        
+        //----------------------------------
+        // Set response
+        //----------------------------------
+        resp.setHPCCAdminUsername(respAdminUser.str());
+
+        VStringBuffer ldapHelm("\n"
+                               "  ldap:\n"
+                               "    adminGroupName: %s\n"
+                               "    hpccAdminSecretKey: TODO!  Create Kubernetes Secret for %s\n"
```

Comment:
Is creating the secret necessary for the respAdminUser to work at all?

Reply:
I cannot create the Kubernetes secret, but it is required for this user to work. Either way the admin will have to create it and specify it in the Helm chart "Secrets: authn: myhpccadmin" and in the ldapenvironment config. Since this is the case, I defer creation of a password to the admin. This "TODO!" is meant to alert the admin they need to do something, perhaps there is a better way?

Reply:
I thought maybe you could provide an option to create the secret if it doesn't exist.  You should be able to do that by calling the kubectl command line tool.  Just like you do for your secrets testing.  We'll have to make sure your pod has the right permissions.  EclWatch can call kubectl for example.